src/org/sosy_lab/common/collect/NaiveOrderStatisticMap.java

```diff
@@ -0,0 +1,162 @@
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.sosy_lab.common.collect;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.collect.ForwardingNavigableMap;
+import com.google.common.collect.Maps;
+import com.google.errorprone.annotations.CanIgnoreReturnValue;
+import java.io.Serializable;
+import java.util.Comparator;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.TreeMap;
+
+/**
+ * An {@link OrderStatisticMap} with naive implementations of its functions.
+ *
+ * <p>The class wraps a {@link NavigableMap} object and delegates all methods inherited from the
+ * <code>NavigableMap</code> interface to that. For the methods particular to the <code>
+ * OrderStatisticMap</code> interface, it provides naive implementations that do not guarantee any
+ * specific performance.
+ *
+ * @param <K> type of the keys of this map. See the Javadoc of {@link OrderStatisticMap} for
+ *     possible constraints on this type
+ * @param <V> type of the values of this map
+ * @see OrderStatisticMap
+ */
+final class NaiveOrderStatisticMap<K, V> extends ForwardingNavigableMap<K, V>
+    implements OrderStatisticMap<K, V>, Serializable {
+
+  private static final long serialVersionUID = -3542217590830996599L;
+
+  private final NavigableMap<K, V> delegate;
+
+  private NaiveOrderStatisticMap(NavigableMap<K, ? extends V> pNavigableMap) {
+    delegate = new TreeMap<>(pNavigableMap);
+  }
+
+  /** Creates a new empty OrderStatisticMap using natural ordering. */
+  static <K, V> NaiveOrderStatisticMap<K, V> createMap() {
+    return new NaiveOrderStatisticMap<>(new TreeMap<>());
+  }
+
+  /** Creates a new empty OrderStatisticMap using the given comparator over its keys. */
+  static <K, V> NaiveOrderStatisticMap<K, V> createMap(Comparator<? super K> pComparator) {
+    return new NaiveOrderStatisticMap<>(new TreeMap<>(checkNotNull(pComparator)));
+  }
+
+  /**
+   * Creates a new OrderStatisticSet containing the same entries as the given map, using natural
+   * ordering over its keys.
+   */
+  static <K, V> NaiveOrderStatisticMap<K, V> createMapWithNaturalOrder(
+      Map<? extends K, ? extends V> pMap) {
+    return new NaiveOrderStatisticMap<>(new TreeMap<>(checkNotNull(pMap)));
+  }
+
+  /**
+   * Creates a new OrderStatisticMap containing the sames entries as the given map and using the
+   * same ordering over its keys as the given map.
+   */
+  static <K, V> NaiveOrderStatisticMap<K, V> createMapWithSameOrder(
+      NavigableMap<K, ? extends V> pNavigableMap) {
+    return new NaiveOrderStatisticMap<>(checkNotNull(pNavigableMap));
+  }
+
+  @Override
+  protected NavigableMap<K, V> delegate() {
+    return delegate;
+  }
+
+  @Override
+  public K getKeyByRank(int pIndex) {
+    return NaiveOrderStatisticSet.createSetWithSameOrder(delegate.navigableKeySet())
```

Comment:
Doesn't this copy the whole set on every access? This seems like a really expensive operation, and most callers of getters will assume that they do not involve large memory allocations. Is there no easy way to avoid this?